src/main/scala/org/apache/spark/shuffle/ucx/NvkvShuffleMapOutputWriter.scala

```diff
@@ -0,0 +1,291 @@
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+//  package org.apache.spark.shuffle.sort.io;
+
+
+object NvkvShuffleMapOutputWriter {
+  private val log = Logger.getLogger("LEO")
+}
+
+class NvkvShuffleMapOutputWriter(private val shuffleId: Int, 
+                                 private val mapId: Long, 
+                                 numPartitions: Int, 
+                                 val ucxTransport: UcxShuffleTransport,
+                                 private val blockResolver: IndexShuffleBlockResolver, 
+                                 sparkConf: SparkConf) extends ShuffleMapOutputWriter {
+  NvkvShuffleMapOutputWriter.log.debug(s"NvkvShuffleMapOutputWriter for shuffleId $shuffleId mapId $mapId numPartitions $numPartitions")
+
+  final private var partitionLengths: Array[Long] = new Array[Long](numPartitions)
+  final private var partitionsPadding: Array[Long] = new Array[Long](numPartitions)
+  private var totalPartitionsPadding: Long = 0
+  final private var bufferSize = 4096
+  private var lastPartitionId = -1
+  private var currChannelPosition = 0L
+  private var dsIdx: Int = 0
+  private var bytesWrittenToMergedFile = 0L
+  private var outputBufferedFileStream: BufferedOutputStream = null
+  private var nvkvWrapper: NvkvWrapper = ucxTransport.getNvkvWrapper
+  private var executerId = SparkEnv.get.blockManager.blockManagerId.executorId.toLong
+
+  private def getBlockOffset = {
+    //TODO - add executer partition offset
+    val numShuffles = 1
+    val numOfMappers = SparkEnv.get.conf.getInt("spark.groupByTest.numMappers", 1)
+    val shuffleBlockSize = nvkvWrapper.getPartitionSize / numShuffles
+    val mapBlockSize = shuffleBlockSize / numOfMappers
+    val r = mapBlockSize % 512;
+    val alignedMapBlockSize = if (r != 0) {mapBlockSize - (r)} else mapBlockSize
+    (shuffleId * shuffleBlockSize) + (mapId * alignedMapBlockSize)
+  }
+
+  @throws[IOException]
+  override def getPartitionWriter(reducePartitionId: Int): ShufflePartitionWriter = {
+    NvkvShuffleMapOutputWriter.log.debug(s"NvkvShuffleMapOutputWriter getPartitionWriter for reducePartition $reducePartitionId")
+    if (reducePartitionId <= lastPartitionId) throw new IllegalArgumentException("Partitions should be requested in increasing order.")
```

Comment:
``
```suggestion
    if (reducePartitionId <= lastPartitionId) {
        throw new IllegalArgumentException("Partitions should be requested in increasing order.")
    }
```
Would also suggest using our own exception types instead of the standard ones
for example:
```scala
class UcxShuffleException extends IllegalArgumentException {
}
```

Reply:
What is the benefit of extending specific Java native exception?
I see reasons why we should not do it:
- It will not help us to better understand the problem.
- I don't think that this exception need any special treatment.

Reply:
I meant something more like a new hierarchy:
```
class UcxShuffleException extends RuntimeException {
}

class UcxShuffleIllegalArgumentException extends UcxShuffleException {
}
```
The benefit is allowing users to distinguish between handling exceptions related to UcxShuffle and other exceptions.
Just a suggestion.

Reply:
IMO UcxShuffle is not really useful hierarchy because it's pretty general and it's encapsulates all the logic more or less. 
If we want to gain some benefit from this approach we can define exception for different instances, for example NvkvWrapper, UcxTransport...
I will do it only if handling this exception requires some special treatment or if we want to handle many exceptions in one try..case block.

I think we can do a better job in handling exceptions just not sure if this approach will really give us any benefits.
@gleon99 WDYT?

Reply:
The general approach of having several exception types organized in a hierarchy which resembles various bad flows - is widely considered as a good and desirable practice.
Also, looking forward for when this project gets more complicated - it is the correct thing to do.

In order to proceed with this PR, I suggest we add this task to our TODO list (@roiedanino) and address it later.